app/questionnaire/answer_store_updater.py

```diff
@@ -0,0 +1,163 @@
+from collections import defaultdict
+
+from app.data_model.answer_store import Answer
+from app.helpers.schema_helpers import get_group_instance_id
+from app.questionnaire.location import Location
+
+
+class AnswerStoreUpdater:
+    """Component responsible for any actions that need to happen as a result of updating the answer store
+    """
+
+    def __init__(self, current_location, schema, questionnaire_store):
+        self._current_location = current_location
+        self._schema = schema
+        self._questionnaire_store = questionnaire_store
+        self._answer_store = self._questionnaire_store.answer_store
+
+    def save_form(self, form):
+        if self._should_save_serialised_answers():
+            self._update_questionnaire_store_with_answer_data(form.serialise())
+        else:
+            self._update_questionnaire_store_with_form_data(form.data)
+
+        self._questionnaire_store.add_or_update()
+
+    def _update_questionnaire_store_with_answer_data(self, answers):
+        survey_answer_ids = self._schema.get_answer_ids_for_block(self._current_location.block_id)
+
+        valid_answers = (
+            answer for answer in answers
+            if answer.answer_id in survey_answer_ids
+        )
+
+        for answer in valid_answers:
+            answer.group_instance_id = get_group_instance_id(self._schema, self._answer_store, self._current_location,
+                                                             answer.answer_instance)
+            self._answer_store.add_or_update(answer)
+
+        if self._current_location not in self._questionnaire_store.completed_blocks:
+            self._questionnaire_store.completed_blocks.append(self._current_location)
```

Comment:
These two lines are duplicated in both `_update_questionnaire_store_with_` methods and could be moved to the `save_form` method.